ARCHITECTURE.md

```diff
@@ -0,0 +1,339 @@
+support this common use case.
+
+## Lumping ACLs and Resources together
+
+Although it currently is only possible to fetch the data contained within a
+Resource in a single request, that Resource _is_ affected by other Resources.
+For example, an Access Control List is defined in a separate Resource, but does
+affect whether and how someone can access that Resource.
+
+With this model, it is not possible to modify both a Resource and the Access
+Control List in an atomic update, that is, it requires two separate HTTP
+requests. This can lead to inconsistent data, e.g. if the network connection
+drops between the two requests and one of them fails.
+
+Although there has been no movement in this direction in the year and a half
+that solid-client has now existed, it could be that the Solid specifications
+at some point get updated to support fetching and updating both a Resource and
+its linked Resources in a single request, e.g. by making the linked Resource
+available in a different RDF Graph (i.e. using the fourth element of an RDF
+Quad). To anticipate such a future, linked Resources are stored in the same
+object as the Resource they are linked from. This enables that at some point, a
+function like `saveSolidDatasetAt` might update not just the data in the
+Resource itself, but also that in one or more linked Resources.
+
+This concern was mainly brought up by @pmcb55, so further questions regarding
+this can be directed to him.
+
+# Concepts
+
+## Resources, Files and SolidDatasets
+
+solid-client refers to everything that can be fetched from a URL as a Resource.
+There are two types of Resources: those containing RDF in a serialisation that
+the server understands ("SolidDatasets"), and those that do not ("Files").
+
+The main difference between the two is that, by virtue of the server
+understanding the structure of the former, we can do partial updates (i.e.
+PATCH requests) and retrieve the data in different formats (currently, the
+spec mandates that servers can serve such data as at least Turtle and JSON-LD).
+
+Thus, data stored as Turtle or JSON-LD can be fetched as a SolidDataset; the
+rest are just treated as regular Files: while might be parsed and manipulated by
+other libraries, solid-client just allows downloads and uploads. This applies to
+binary file types like JPEG, WebM and `.txt`, but also to files containing
+structured data in a non-RDF format, like JSON, XML or OpenDocument, or even
+to files containing structured data in an RDF format not understood by the
+server, such as RDFa or RDF-XML.
+
+## `With*` types
+
+Although SolidDatasets and Files (see previous section) are different types of
+Resources, there is certain data that is common to all Resources. This data is
+set on properties for either type of Resource, which is specified in a `With*`
+type.
+
+For example, there is some metadata determining where a particular Resource is
+fetched from; this is set on the property `internal_resourceInfo` for both
```

Comment:
Since we mention the method name explicitly, would it be worth adding a direct link to the file where this is defined ?

Reply:
I'm not sure which method you're referring to (there's no mention of a method in the linked lines), but since they might move around in the future and are relatively easy to find yourself, I tried to avoid that.

Reply:
Sorry I meant the property name. And right, it's true it's easy enough to find it.